### PR TITLE
LC30CHALLENGE: Problem №10

### PR DESCRIPTION
**Pull request by DmitryBokovikov**

### CODE REVIEW

File: src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem10.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+{
+template <class T>
+class MinStackImpl {
+ public:
+  void push(const T& x) {
+    if (empty(m_data)) {
+      m_data.emplace_back(x, x);
+    }
+    else {
+      auto[_, previous_min] = m_data.back();
```
Comment (fbokovikov) [inline src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem10.h:13]: this branch may be simplified to:
```
m_data.emplace_back(x, std::min(x, getMin()));
```

File: src/main/cplusplus/leetcode_challenge/leetcode_challenge/problems.h [hunk 2]
```diff
@@ -31,4 +31,6 @@ namespace problems_ns
 
     // day 9
     bool backspaceCompare(std::string_view s1, std::string_view s2);
+
+    // day 10 implement min stack
```
Comment (fbokovikov) [inline src/main/cplusplus/leetcode_challenge/leetcode_challenge/problems.h:35]: did you forget to add header?

Comment (DmitryBokovikov) [reply src/main/cplusplus/leetcode_challenge/leetcode_challenge/problems.h:35]: I guess you meant declaration not header. Anyway the answer is no - it was header implemented so the only thing I could do is to mention solution header in problems.h

File: src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem10.h (hunk 1 above)
Comment (DmitryBokovikov) [reply src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem10.h:13]: Did you get acquinted  with emplace semantics? How do you find it?

Comment (DmitryBokovikov) [reply src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem10.h:13]: thanks, good point. But it's not the reason to decline PR :)

File: src/main/cplusplus/leetcode_challenge/leetcode_challenge/problems.h [hunk 3]
```diff
@@ -31,4 +31,9 @@ namespace problems_ns
...
+
+    // day 11
+    int diameterOfBinaryTree(struct TreeNode* root);
```
Comment (DmitryBokovikov) [inline src/main/cplusplus/leetcode_challenge/leetcode_challenge/problems.h:38]: It's like to emphasize that this header knows nothing about TreeNode structure. But it uses it in declaration. This type of declaration of TreeNode is called forward declaration. JFYI.

File: src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem10.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  int getMin() {
+    return m_data.back().second;
+  }
+ private:
+  std::vector<std::pair<T, T>> m_data;
+};
+
+using MinStack = MinStackImpl<int>;
```
Comment (DmitryBokovikov) [inline src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem10.h:32]: using directive is the same thing as typealias in Kotlin, JFYI.

File: src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem11.cpp [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+int diameterOfBinaryTree(TreeNode* root) {
+  std::function<int(TreeNode* p)> processNode;
+  int ans = 0;
+  processNode = [&](TreeNode* p) {
+    if (!p) {
+      return 0;
+    }
+    const auto left = processNode(p->left);
+    const auto right = processNode(p->right);
+    ans = std::max(ans, left + right + 1);
```
Comment (fbokovikov) [inline src/main/cplusplus/leetcode_challenge/leetcode_challenge/problem11.cpp:14]:
```
ans = std::max(ans, left + right);
```
and 
```
return ans;
```
What's the point to add 1 and then subtract 1?

**Pull request merged**